File: haskell/p205.hs
{- 
 - Solution to Project Euler problem 205
 - Copyright (c) Project Nayuki. All rights reserved.
 - 
 - https://www.nayuki.io/page/project-euler-solutions
 - https://github.com/nayuki/Project-Euler-solutions
 -}

import Data.Ratio ((%))
import Numeric (showFFloat)


main = putStrLn (ans "")
numer = sum [(ninePyramidalPdf !! i) * (sum (take i sixCubicPdf)) | i <- [0 .. ((length ninePyramidalPdf) - 1)]]
denom = (sum ninePyramidalPdf) * (sum sixCubicPdf)
ans = showFFloat (Just 7) (fromRational (numer % denom))


pyramidalDiePdf = [0, 1, 1, 1, 1] :: [Integer]
ninePyramidalPdf = (iterate (convolve pyramidalDiePdf) [1]) !! 9

cubicDiePdf = [0, 1, 1, 1, 1, 1, 1] :: [Integer]
sixCubicPdf = (iterate (convolve cubicDiePdf) [1]) !! 6


convolve :: [Integer] -> [Integer] -> [Integer]
convolve as bs = let
		m = length as
		n = length bs
		func i = sum [(as !! j) * (bs !! (i - j)) | j <- [(max 0 (i - (n - 1))) .. (min i (m - 1))]]
	in map func [0 .. (m + n - 2)]

File: python/p205.py
# 
# Solution to Project Euler problem 205
# Copyright (c) Project Nayuki. All rights reserved.
# 
# https://www.nayuki.io/page/project-euler-solutions
# https://github.com/nayuki/Project-Euler-solutions
# 


def compute():
	nine_pyramidal_pdf = [1]
	PYRAMIDAL_DIE_PDF = [0, 1, 1, 1, 1]
	for i in range(9):
		nine_pyramidal_pdf = convolve(nine_pyramidal_pdf, PYRAMIDAL_DIE_PDF)
	
	six_cubic_pdf = [1]
	CUBIC_DIE_PDF = [0, 1, 1, 1, 1, 1, 1]
	for i in range(6):
		six_cubic_pdf = convolve(six_cubic_pdf, CUBIC_DIE_PDF)
	
	ans = 0
	for i in range(len(nine_pyramidal_pdf)):
		ans += nine_pyramidal_pdf[i] * sum(six_cubic_pdf[ : i])
	ans = float(ans) / (sum(nine_pyramidal_pdf) * sum(six_cubic_pdf))
	return f"{ans:.7f}"


def convolve(a, b):
	c = [0] * (len(a) + len(b) - 1)
	for i in range(len(a)):
		for j in range(len(b)):
			c[i + j] += a[i] * b[j]
	return c


if __name__ == "__main__":
	print(compute())
